File: app/src/main/java/ch/unstable/ost/BaseNavigationFragment.java
package ch.unstable.ost;

import android.content.Context;


import androidx.annotation.CallSuper;
import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import ch.unstable.ost.api.model.ConnectionQuery;


public abstract class BaseNavigationFragment extends Fragment {
    private OnRouteSelectionListener mListener;

    @CallSuper
    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnRouteSelectionListener) {
            mListener = (OnRouteSelectionListener) context;
        } else {
            throw new IllegalStateException(context.toString()
                    + " must implement OnRouteSelectionListener");
        }
    }

    @CallSuper
    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public void selectRoute(ConnectionQuery query) {
        if (mListener != null) {
            mListener.onRouteSelected(query);
        }
    }

    public interface OnRouteSelectionListener {

        @MainThread
        void onRouteSelected(@NonNull ConnectionQuery query);
    }
}

File: app/src/main/java/ch/unstable/ost/BaseNavigationFragment.kt
package ch.unstable.ost

import android.content.Context
import androidx.annotation.CallSuper
import androidx.annotation.MainThread
import androidx.fragment.app.Fragment
import ch.unstable.ost.api.model.ConnectionQuery

abstract class BaseNavigationFragment : Fragment() {
    private var listener: OnRouteSelectionListener? = null
    @CallSuper
    override fun onAttach(context: Context) {
        super.onAttach(context)
        listener = if (context is OnRouteSelectionListener) {
            context
        } else {
            throw IllegalStateException(context.toString()
                    + " must implement OnRouteSelectionListener")
        }
    }

    @CallSuper
    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    fun selectRoute(query: ConnectionQuery) {
        listener?.onRouteSelected(query)
    }

    interface OnRouteSelectionListener {
        @MainThread
        fun onRouteSelected(query: ConnectionQuery)
    }
}